[Migration guide] Improve part about using translations in AppProvider (#2312)
Shopify/polaris-react: src/components/AppProvider/README.md

Translations are provided in the locales folder. When using Polaris, you are able to import translations from all languages supported by the core Shopify product and consume them through the `i18n` prop.

> At its simplest you can pass the json content from a locale file into `AppProvider`. This approach is suitable when a project has only one specific locale.

Reviewer: Can we simplify to something like this?

```suggestion
If a project has only one locale, then you can pass the JSON content from the locale file into `AppProvider`. 
```

Reviewer 2: Much nicer! And thank you for getting rid of my use of "simplest". Describing something as simple is a bold assumption and I should know better :D

```jsx
import AppProvider from '@shopify/polaris';
// en.json is English. Replace with fr.json for French, etc
import translations from '@shopify/polaris/locales/en.json';

function App() {
  return <AppProvider i18n={translations}>{/* App content */}</AppProvider>;
}
```

> Alternatively if your project supports multiple locales you can load them dynamically using [@shopify/react-i18n](https://github.com/Shopify/quilt/tree/master/packages/react-i18n#translation). This will ensure you only load the translations you need instead of potentially loading every set of translations even if you are not using them.

Reviewer: ```suggestion
If a project supports multiple locales, then load them dynamically using [@shopify/react-i18n](https://github.com/Shopify/quilt/tree/master/packages/react-i18n#translation). This ensures that you load only the translations you need.
```

```jsx
import AppProvider from '@shopify/polaris';
// en.json is English. Replace with fr.json for French, etc
import translations from '@shopify/polaris/locales/en.json';
import {useI18n} from '@shopify/react-i18n';

function App() {
  const [i18n] = useI18n({
    id: 'Polaris',
    fallback: translations,
    translations(locale) {
      return import(

>         /* webpackChunkName: "Polaris-i18n-[request]", webpackMode: "lazy-once" */ `@shopify/polaris/locales/${locale}.json`

Reviewer 3: I'm not too familiar with `react-i18n`, does this require specific webpack configuration?

Reviewer 2: Nothing beyond what sewing-kit already provides

Reviewer 3: We may want to add a section here about webpack being required

Reviewer 2: That seems pretty strongly implied. This section is targeted at shopify developers, it's almost guaranteed that they'll be working on a project that uses sewing-kit, and thus webpack and react-i18n will be already preconfigured anyway.

      ).then((dictionary) => dictionary && dictionary.default);
    },
  });

>   // i18n.translations is an array of translation dictionaries, where the first

Reviewer 3: `translation dictionaries` is a named for a type we've used but consumers may not know what we're referring too. What do you think about describing it the type?

Reviewer 2: That feels like an implementation detail of react-i18n, I don't think we need to go into much detail other than "here's what useI18n gives us, here's how we need to tweak it so it works with our internationalisation".

Reviewer 3: iirc i18n.translations can include various translations. If we're wanting people to use `react-i18n`, they'll want to use it for their custom translations as well rather than have two i18n dependencies. i18n.translations could be [polaris_fallback, polaris_jp, custom translations]. I'm not sure how much detail we'll want to include, but it might help the consumer become more aware of what's actually happening.

Reviewer 2: I don't think it's our responsibility to describe how react-i18n works here - that's what their docs are for. This is true for the common case that we recommend and if people want to use a more advanced setup then it's up to them to work out how they want to diverge from our recommendations.

Translations are provided in the locales folder. When using Polaris, you are able to import translations from all languages supported by the core Shopify product and consume them through the `i18n` prop.

> ```jsx

Reviewer 2: I think referencing a `translations` variable without showing where it comes from is confusing. Let's keep this as an example of the static translations, then segue into dynamic translations - that way we might not need the explicit headings.

How does something like this sound? (Sorry for only coming up with this after you wrote the first draft):

> Translations are provided in the locales folder. When using Polaris, you are able to import translations from all languages supported by the core Shopify product and consume them through the `i18n` prop.
> 
> At its simplest you can pass the json content from a locale file into `AppProvider`. This approach is suitable when a project has only one specific locale.
> 
> ```jsx
> import AppProvider from "@shopify/polaris";
> // en.json is English. Replace with fr.json for French, etc
> import translations from "@shopify/polaris/locales/en.json";
> 
> function App() {
>   return <AppProvider i18n={translations}>{/* App content */}</AppProvider>;
> }
> ```
> 
> Alternatively if your project supports multiple locales you can load them dynamically using [`@shopify/react-i18n`](https://github.com/Shopify/quilt/tree/master/packages/react-i18n#translation). This will ensure you only load the translations you need instead of potentially loading every set of translations even if you are not using them.
> 
> ```jsx
> import AppProvider from "@shopify/polaris";
> // en.json is English. Replace with fr.json for French, etc
> import translations from "@shopify/polaris/locales/en.json";
> import { useI18n } from "@shopify/react-i18n";
> 
> function App() {
>   const [i18n] = useI18n({
>     id: "Polaris",
>     fallback: translations,
>     translations(locale) {
>       return import(
>         /* webpackChunkName: "Polaris-i18n-[request]", webpackMode: "lazy-once" */ `@shopify/polaris/locales/${locale}.json`
>       ).then(dictionary => dictionary && dictionary.default);
>     }
>   });
> 
>   // i18n.translations is an array of translation dictionaries, where the first
>   // dictionary is the desired language, and the second is the fallback.
>   // AppProvider however expects that the first dictionary is the fallback
>   // and the second is the desired language. Thus we need to reverse the array
>   // to ensure the dictionaries are in the order desired by AppProvider
>   return (
>     <AppProvider i18n={i18n.translations.reverse()}>{/* App content */}</AppProvider>
>   );
> }
> ```

---

I've modified the translations function there, but I've not ran it to test as I've not got an environment setup for that. Would you be able to check it for me - it is based up on a simplified version of the code the [react-i18n babel plugin](https://github.com/Shopify/quilt/blob/3f8f237de601f93501f6a2ad0d63bdbe845ea43c/packages/react-i18n/src/babel-plugin.ts#L30-L48) generates - that's where the "webpacMode" comes from. 

I've removed any typescript types as we want our examples to be plain JS. I've also removed the `const translationsLocale = locale && locale !== '' ? locale : 'en';` check as [react-i18n's types](https://github.com/Shopify/quilt/blob/3f8f237de601f93501f6a2ad0d63bdbe845ea43c/packages/react-i18n/src/manager.ts#L9-L11) suggests that a string will always be passed into the translate function. I've also added a comment explaining why the reverse is needed inline, rather than explaining it in the docs (related - we should fix the ordering to match usei18n in the next major version to avoid this fiddling)

Author: agree that fixing the order will help - then there will be no need to understand this reverse "gotcha"
